hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImplV2.java

```diff
@@ -0,0 +1,398 @@
+        long nextTXID = UniqueId.next();
+        DeletedBlocksTransaction tx = constructNewTransaction(nextTXID,
+            entry.getKey(), entry.getValue());
+        txsToBeAdded.add(tx);
+      }
+
+      deletedBlockLogStateManager.addTransactionsToDB(txsToBeAdded);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+  }
+
+  private void getTransaction(DeletedBlocksTransaction tx,
+      DatanodeDeletedBlockTransactions transactions) {
+    try {
+      Set<ContainerReplica> replicas = containerManager
+          .getContainerReplicas(ContainerID.valueOf(tx.getContainerID()));
+      for (ContainerReplica replica : replicas) {
+        UUID dnID = replica.getDatanodeDetails().getUuid();
+        Set<UUID> dnsWithTransactionCommitted =
+            transactionToDNsCommitMap.get(tx.getTxID());
+        if (dnsWithTransactionCommitted == null || !dnsWithTransactionCommitted
+            .contains(dnID)) {
+          // Transaction need not be sent to dns which have
+          // already committed it
+          transactions.addTransactionToDN(dnID, tx);
+        }
+      }
+    } catch (IOException e) {
+      LOG.warn("Got container info error.", e);
+    }
+  }
+
+  @Override
+  public DatanodeDeletedBlockTransactions getTransactions(
+      int blockDeletionLimit) throws IOException {
+    lock.lock();
+    try {
+      DatanodeDeletedBlockTransactions transactions =
+          new DatanodeDeletedBlockTransactions();
+      try (TableIterator<Long,
+          ? extends Table.KeyValue<Long, DeletedBlocksTransaction>> iter =
+               deletedBlockLogStateManager.getReadOnlyIterator()) {
+        int numBlocksAdded = 0;
+        ArrayList<Long> txIDs = new ArrayList<>();
+        while (iter.hasNext() && numBlocksAdded < blockDeletionLimit) {
+          Table.KeyValue<Long, DeletedBlocksTransaction> keyValue = iter.next();
+          DeletedBlocksTransaction txn = keyValue.getValue();
+          final ContainerID id = ContainerID.valueOf(txn.getContainerID());
+          try {
+            if (txn.getCount() > -1 && txn.getCount() <= maxRetry
+                && !containerManager.getContainer(id).isOpen()) {
+              numBlocksAdded += txn.getLocalIDCount();
+              getTransaction(txn, transactions);
+              transactionToDNsCommitMap
+                  .putIfAbsent(txn.getTxID(), new LinkedHashSet<>());
+            }
+          } catch (ContainerNotFoundException ex) {
+            LOG.warn("Container: " + id + " was not found for the transaction: "
+                + txn);
+            txIDs.add(txn.getTxID());
+          }
+        }
+
+        deletedBlockLogStateManager.removeTransactionsFromDB(txIDs);
+      }
+      return transactions;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void onMessage(
+      DeleteBlockStatus deleteBlockStatus, EventPublisher publisher) {
+    if (!scmContext.isLeader()) {
```

Comment:
I am a bit confused here: so a delete block message could be still in-flight, and before it is processed, SCM leader could change, thus when `onMessage` is called, we can check whether `scmContext.isleader()` to confirm whether to proceed?

Is this  `scmContext.isleader()` just an optimization? E.g. because leader change can still happen after this check, thus those Ratis operations will fail by a NotLeaderException?

Reply:
It is just an optimization, if without this check, when leader change, it only waste a ratis call which will fail

Reply:
Got it. Thanks for clarification.